### PR TITLE
Airflow Agent Refactor

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
- Use imageSpec inside the airflow operator
- Run dataflow operator in the container
- Remove hacky code
- Remove google dependencies

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Here is an example
```python
import os

from airflow.providers.apache.beam.operators.beam import BeamRunPythonPipelineOperator, BeamRunJavaPipelineOperator
from airflow.providers.google.cloud.operators.dataflow import DataflowConfiguration
from click.testing import CliRunner

from flytekit import workflow, ImageSpec
from flytekit.clis.sdk_in_container import pyflyte

new_airflow_plugin = "git+https://github.com/flyteorg/flytekit.git@8b41e63fa50070bbdc0a3486763ceafec892f932#subdirectory=plugins/flytekit-airflow"
image_spec = ImageSpec(
    registry="pingsutw",
    base_image="pingsutw/flytekit:dataflow-v2",
    packages=["apache-airflow[google]", "apache-airflow-providers-apache-beam[google]", new_airflow_plugin],
    apt_packages=["git"],
    env={"AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT": "google-cloud-platform://"}
)


@workflow
def wf():
    BeamRunJavaPipelineOperator(
        task_id="start_java_pipeline_dataflow_runner",
        runner="DataflowRunner",
        job_class="org.apache.beam.examples.WordCount",
        pipeline_options={"output": "gs://opta-gcp-dogfood-gcp/spark",
                          "stagingLocation": "gs://opta-gcp-dogfood-gcp/staging",
                          "tempLocation": "gs://opta-gcp-dogfood-gcp/staging1",
                          "region": "us-east1"},
        jar="gs://opta-gcp-dogfood-gcp/spark/word-count-beam-bundled-0.1.jar",
        dataflow_config=DataflowConfiguration(job_name="test_dataflow",
                                              project_id="dogfood-gcp-dataplane",
                                              location="us-east1",
                                              wait_until_finished=False),
        container_image=image_spec
    )

    BeamRunPythonPipelineOperator(
        task_id="start_python_pipeline_dataflow_runner",
        runner="DataflowRunner",
        py_file="gs://opta-gcp-dogfood-gcp/spark/word_count.py",
        pipeline_options={"output": "gs://opta-gcp-dogfood-gcp/spark",
                          "input": "gs://opta-gcp-dogfood-gcp/spark/-00000-of-00001",
                          "stagingLocation": "gs://opta-gcp-dogfood-gcp/staging1",
                          "region": "us-east1",
                          "tempLocation": "gs://opta-gcp-dogfood-gcp/staging1"},
        py_interpreter="python3",
        dataflow_config=DataflowConfiguration(job_name="test_dataflow", project_id="dogfood-gcp-dataplane",
                                              location="us-east1",
                                              wait_until_finished=False),
        container_image=image_spec
    )


if __name__ == '__main__':
    os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = "/Users/kevin/git/flytekit/somekey.json"
    os.environ['AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT'] = "google-cloud-platform://"
    os.environ['FLYTECTL_CONFIG'] = '/Users/kevin/.flyte/config-dogfood-gcp.yaml'
    runner = CliRunner()
    result = runner.invoke(pyflyte.main,
                           ["run", "--remote", "/Users/kevin/git/flytekit/flyte-example/airflow_example/dataflow_java.py",
                            "wf"])
    print(result.output)
```

Base image:
```Dockerfile
ARG PYTHON_VERSION
FROM python:${PYTHON_VERSION}-slim-buster

MAINTAINER Flyte Team <users@flyte.org>
LABEL org.opencontainers.image.source=https://github.com/flyteorg/flytekit

WORKDIR /root
ENV PYTHONPATH /root

ARG VERSION
ARG DOCKER_IMAGE

RUN apt-get update && apt-get install build-essential -y

# Pod tasks should be exposed in the default image
RUN pip install -U flytekit==$VERSION \
	flytekitplugins-pod==$VERSION \
	flytekitplugins-deck-standard==$VERSION \
	scikit-learn

RUN apt-get update && \
    apt-get install -y curl gnupg && \
    echo "deb [signed-by=/usr/share/keyrings/cloud.google.gpg] http://packages.cloud.google.com/apt cloud-sdk main" | tee -a /etc/apt/sources.list.d/google-cloud-sdk.list && \
    curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key --keyring /usr/share/keyrings/cloud.google.gpg  add - && \
    apt-get update -y && \
    apt-get install google-cloud-sdk -y

RUN mkdir -p /usr/share/man/man1 /usr/share/man/man2 && \
    apt-get update &&\
    apt-get install -y --no-install-recommends openjdk-11-jre && \
    apt-get install ca-certificates-java -y && \
    apt-get clean && \
    update-ca-certificates -f;
ENV JAVA_HOME /usr/lib/jvm/java-11-openjdk-amd64/


RUN useradd -u 1000 flytekit
RUN chown flytekit: /root
# Some packages will create config file under /home by default, so we need to make sure it's writable
RUN chown flytekit: /home
USER flytekit

ENV FLYTE_INTERNAL_IMAGE "$DOCKER_IMAGE"

```

image

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3936

## Follow-up issue
_NA_

### CODE REVIEW

File: plugins/flytekit-airflow/flytekitplugins/airflow/task.py [hunk 1]
```diff
@@ -1,51 +1,175 @@
...
+    # airflow task instead of the Flyte task.
+    ctx = FlyteContextManager.current_context()
+    ctx.user_space_params.builder().add_attr("GET_ORIGINAL_TASK", True).build()
+
+    obj_module = importlib.import_module(name=airflow_obj.module)
+    obj_def = getattr(obj_module, airflow_obj.name)
+    if (
+        issubclass(obj_def, BaseOperator)
+        and not issubclass(obj_def, BaseSensorOperator)
+        and not _is_dataflow_operator(obj_def)
```
Comment (kumare3) [inline plugins/flytekit-airflow/flytekitplugins/airflow/task.py:148]:
```suggestion
        and not _is_not_supported(obj_def)
```

File: plugins/flytekit-airflow/flytekitplugins/airflow/task.py [hunk 2]
```diff
@@ -1,51 +1,175 @@
...
+    ):
+        try:
+            return obj_def(**airflow_obj.parameters, deferrable=True)
+        except AirflowException:
+            logger.debug(f"Airflow operator {airflow_obj.name} does not support deferring")
+
+    return obj_def(**airflow_obj.parameters)
+
+
+def _is_dataflow_operator(cls: Type):
```
Comment (kumare3) [inline plugins/flytekit-airflow/flytekitplugins/airflow/task.py:158]:
```suggestion
def _is_not_supported(cls: Type):
```
Or can be called `_is_not_agent_supported`

File: plugins/flytekit-airflow/flytekitplugins/airflow/task.py [hunk 3]
```diff
@@ -1,51 +1,175 @@
...
+    """
+    This function is used to check if task is a dataflow operator.
+    """
+    try:
```
Comment (kumare3) [inline plugins/flytekit-airflow/flytekitplugins/airflow/task.py:162]: this this can be multiple if-else

File: plugins/flytekit-airflow/flytekitplugins/airflow/task.py [hunk 4]
```diff
@@ -57,10 +181,18 @@ def _flyte_operator(*args, **kwargs):
...
         logging.debug("failed to get the attribute GET_ORIGINAL_TASK from user space params")
-    config = AirflowConfig(task_module=cls.__module__, task_name=cls.__name__, task_config=kwargs)
-    t = AirflowTask(name=kwargs["task_id"], query_template="", task_config=config, original_new=cls.__new__)
-    return t()
+
+    container_image = kwargs.pop("container_image", None)
+    task_id = kwargs["task_id"] or cls.__name__
+    config = AirflowObj(module=cls.__module__, name=cls.__name__, parameters=kwargs)
+
+    if _is_dataflow_operator(cls):
```
Comment (kumare3) [inline plugins/flytekit-airflow/flytekitplugins/airflow/task.py:192]: same thing

File: Makefile [hunk 5]
```diff
@@ -35,9 +35,9 @@ fmt:
 lint: ## Run linters
 	mypy flytekit/core
 	mypy flytekit/types
-	# allow-empty-bodies: Allow empty body in function.
-	# disable-error-code="annotation-unchecked": Remove the warning "By default the bodies of untyped functions are not checked".
-	# Mypy raises a warning because it cannot determine the type from the dataclass, despite we specified the type in the dataclass.
+#	allow-empty-bodies: Allow empty body in function.
+#	disable-error-code="annotation-unchecked": Remove the warning "By default the bodies of untyped functions are not checked".
+#	Mypy raises a warning because it cannot determine the type from the dataclass, despite we specified the type in the dataclass.
```
Comment (eapolinario) [inline Makefile:40]: nit: can you revert these changes?

File: plugins/flytekit-airflow/flytekitplugins/airflow/agent.py [hunk 6]
```diff
@@ -1,108 +1,145 @@
...
+                    try:
+                        # Trigger callback will check the status of the task in the payload, and raise AirflowException if failed.
+                        trigger_callback = getattr(airflow_operator_instance, meta.airflow_trigger_callback)
+                        trigger_callback(context=airflow_ctx, event=typing.cast(TriggerEvent, event).payload)
+                        cur_state = SUCCEEDED
+                    except AirflowException as e:
+                        cur_state = RETRYABLE_FAILURE
+                        message = e.__str__()
+                except asyncio.TimeoutError:
+                    logger.debug("No event received from airflow trigger")
```
Comment (eapolinario) [inline plugins/flytekit-airflow/flytekitplugins/airflow/agent.py:126]: why we don't capture this timeout in `cur_state`?

File: plugins/flytekit-airflow/flytekitplugins/airflow/task.py [hunk 7]
```diff
@@ -1,51 +1,171 @@
...
+    ctx = FlyteContextManager.current_context()
+    ctx.user_space_params.builder().add_attr("GET_ORIGINAL_TASK", True).build()
+
+    obj_module = importlib.import_module(name=airflow_obj.module)
+    obj_def = getattr(obj_module, airflow_obj.name)
+    if issubclass(obj_def, BaseOperator) and not issubclass(obj_def, BaseSensorOperator) and _is_deferrable(obj_def):
+        try:
+            return obj_def(**airflow_obj.parameters, deferrable=True)
+        except AirflowException:
+            logger.debug(f"Airflow operator {airflow_obj.name} does not support deferring")
```
Comment (eapolinario) [inline plugins/flytekit-airflow/flytekitplugins/airflow/task.py:149]: do we have to return here?

File: Makefile (hunk 5 above)
Comment (pingsutw) [reply Makefile:40]: This is intended. if we revert it, it will always show on the terminal.
```bash
(flytekit-3.10) ➜  flytekit git:(error-handler) ✗ make lint
mypy flytekit/core
flytekit/core/workflow.py:463: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
Success: no issues found in 43 source files
mypy flytekit/types
flytekit/types/structured/structured_dataset.py:303: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
Success: no issues found in 21 source files
# allow-empty-bodies: Allow empty body in function.
# disable-error-code="annotation-unchecked": Remove the warning "By default the bodies of untyped functions are not checked".
# Mypy raises a warning because it cannot determine the type from the dataclass, despite we specified the type in the dataclass.
mypy --allow-empty-bodies --disable-error-code="annotation-unchecked" tests/flytekit/unit/core
Success: no issues found in 75 source files
pre-commit run --all-files
```

File: plugins/flytekit-airflow/flytekitplugins/airflow/agent.py (hunk 6 above)
Comment (pingsutw) [reply plugins/flytekit-airflow/flytekitplugins/airflow/agent.py:126]: if no event received from airflow trigger, it means job still running. Default state here is running

File: plugins/flytekit-airflow/flytekitplugins/airflow/task.py (hunk 7 above)
Comment (pingsutw) [reply plugins/flytekit-airflow/flytekitplugins/airflow/task.py:149]: yes, we return at line 151 without `deferrable=True`

Comment (eapolinario) [reply plugins/flytekit-airflow/flytekitplugins/airflow/task.py:149]: got it. 

Another question: is there a more specific exception related to operators not being deferrable? Capturing `AirflowException` means that other exceptions are also treated as that, right? If that's not the case, can we get the exception message in the logs as well?

Comment (pingsutw) [reply plugins/flytekit-airflow/flytekitplugins/airflow/task.py:149]: It always raises AirflowException
https://github.com/apache/airflow/blob/3a084b70c2161a02a0cb0db39bd06dfb120de46d/airflow/models/baseoperator.py#L808-L811

Comment (pingsutw) [reply plugins/flytekit-airflow/flytekitplugins/airflow/task.py]: updated error message.

**Review by eapolinario — APPROVED** [review]
